File: Back end/packages/entity/api/src/main/java/vn/aeoc/packages/entity/api/Product.java
package vn.aeoc.packages.entity.api;

import jakarta.validation.constraints.NotEmpty;
//import jakarta.validation.constraints.NotNull;
import lombok.Data;

@Data
public class Product {

    private Integer id;
    @NotEmpty(message = "Tên sản phẩm không để trống")
    private String name;
    private String description;
//    @NotNull(message = "Danh mục sản phẩm không để trống")
    private Integer categoryId;
    private String categoryName;
    private Float avgRating = 0f;
    private String image;
    private String thumbnail;
    private Integer active;
    private Double minPrice = 0.0;
    private Double maxPrice = 0.0;
    private Double originalMinPrice = 0.0;
    private Double originalMaxPrice = 0.0;

    public Product(Product product) {
        this.id = product.id;
        this.name = product.name;
        this.description = product.description;
        this.categoryId = product.categoryId;
        this.categoryName = product.categoryName;
        this.avgRating = product.avgRating;
        this.image = product.image;
        this.thumbnail = product.thumbnail;
        this.active = product.active;
        this.minPrice = product.minPrice;
        this.maxPrice = product.maxPrice;
        this.originalMinPrice = product.originalMinPrice;
        this.originalMaxPrice = product.originalMaxPrice;
    }

    public Product() {}
}

File: Front end/src/types/product.ts
export interface AttributeValue {
  id: number
  attributeId: number
  value: string
}

export interface ProductAttribute {
  id?: number
  productId: number
  fieldName: string
  name: string
  createdAt?: string
  updatedAt?: string
  options: AttributeValue[]
}

export interface ProductVariant {
  id?: number
  productId: number
  productName?: string
  sku: string
  image: string
  price: number
  originalPrice?: number
  stock: number
  attributeValueIds: number[]
  options?: Record<string, string>
  createdAt?: string
  updatedAt?: string
}

export interface Product {
  id: number
  name: string
  description: string
  categoryId: number
  categoryName?: string
  avgRating: number
  image: string
  thumbnail: string
  active: number
  minPrice: number
  maxPrice: number
  originalMinPrice?: number
  originalMaxPrice?: number
}

export interface ProductDetail extends Product {
  attributes: ProductAttribute[]
  variants: ProductVariant[]
}

export interface ProductListResponse {
  total: number
  page: number
  items: Product[]
}

export interface ProductDetailListResponse {
  total: number
  page: number
  items: ProductDetail[]
}

export interface GetProductsRequest {
  keyword?: string
  categoryId?: number
  active?: number
  priceFrom?: number
  priceTo?: number
  avgRatingFrom?: number
  avgRatingTo?: number
  sortBy?: string
  sortDir?: 'asc' | 'desc'
  page?: number
  limit?: number
  ids?: number[]
}

export interface GetProductDetailRequest {
  id: number | string
}

export interface ProductFormData {
  id?: number
  name: string
  description: string
  categoryId: number
  image: string
  thumbnail: string
  active: number
}

export interface Category {
  id: number
  name: string
  description?: string
}
